Added some limelight and trajectory code with some controls.

LoganR-FRC-Team34 opened the pull request:
Added some limelight and trajectory code with some controls commented some stuff out for moving the arm and the intake because currently there is no infrared sensor, and the arm encoders are not returning values.

LoganR-FRC-Team34 committed:
Added some limelight and trajectory code with some controls. Commented some stuff out for moving the arm and the intake because currently there is no infrared sensor and the arm encoders are not returning values.

bradleyscollins commented on src/main/include/subsystems/Shooter.h:
```diff
@@ -40,14 +42,18 @@ namespace t34
-        inline double GetBottomArmEncoderVal() { return m_arm_encoder_top.GetPosition(); }
+        inline double GetBottomArmEncoderVal() { return m_arm_encoder_bottom.GetPosition(); }
 
         inline void RunTopArmMotor(const double motor_output) { m_arm_motor_top.Set(motor_output); }
         inline void RunBottomArmMotor(const double motor_output) { m_arm_motor_bottom.Set(motor_output); }
```
When these functions are called in other places, it is difficult to know what they mean. E.g., ...

```cpp
rc->shooter.RunTopArmMotor(0.4);
rc->shooter.RunBottomArmMotor(0.4);
```

Is that 0.4 a fraction/percentage value? If so, it might be clearer to include that in the function name:

```suggestion
        inline void RunTopArmMotorPercent(const double percent) { m_arm_motor_top.Set(percent); }
        inline void RunBottomArmMotorPercent(const double percent) { m_arm_motor_bottom.Set(percent); }
```

bradleyscollins commented on src/main/include/subsystems/Shooter.h:
```diff
@@ -40,14 +42,18 @@ namespace t34
 
         void RunIntake(const double motor_output);
```
As with the suggestion about `RunTopArmMotor()`, these might benefit from a similar convention:

```suggestion
        void RunShooter(const double percent);
        void RunIntake(const double percent);
```

bradleyscollins commented on src/main/include/subsystems/LimeLightUtil.h:
```diff
@@ -0,0 +1,61 @@
+
+        double m_target_height_meters;
+        double m_note_velocity_mps;
+        double m_ll_angle_deg;
+        double m_target_distance_meters;
+        double m_ll_height_meters;
```
Can I be a stickler about spelling things out?

```suggestion
        double m_limelight_angle_deg;
        double m_target_distance_meters;
        double m_limelight_height_meters;
```

bradleyscollins commented on src/main/include/subsystems/LimeLightUtil.h:
```diff
@@ -0,0 +1,61 @@
+namespace t34
+{
+
+    class LimeLightUtil
```
The [Limelight folks](https://limelightvision.io/) don't capitalize the L in light. Suggest following their convention:

```suggestion
    class LimelightUtil
```

bradleyscollins commented on src/main/include/subsystems/LimeLightUtil.h:
```diff
@@ -0,0 +1,61 @@
+
+    private:
+
+        frc::PIDController m_ll_swerve_pid;
```
May I be a stickler about spelling things out?

```suggestion
        frc::PIDController m_limelight_swerve_pid;
```

bradleyscollins commented on src/main/include/subsystems/LimeLightUtil.h:
```diff
@@ -0,0 +1,61 @@
+#pragma once
+
+#include <frc/controller/PIDController.h>
+#include "LimeLightHelpers.h"
```
The [Limelight folks](https://limelightvision.io/) don't capitalize the L in light. Suggest following their lead:

```suggestion
#include "LimelightHelpers.h"
```

bradleyscollins commented on src/main/include/Constants.h:
```diff
@@ -11,6 +12,8 @@ constexpr double NEO550_RES{ 42 };
 constexpr double INTAKE_GEAR_RATIO{203.636364};
 constexpr double ARM_ENC_CONVERSION_FACTOR{360.0 / (NEO550_RES * INTAKE_GEAR_RATIO)};
 
+const std::string LL_TABLE_NAME{ "" };
```
May I be a stickler about spelling things out?

```suggestion
const std::string LIMELIGHT_TABLE_NAME{ "" };
```

bradleyscollins commented on src/main/cpp/Traj.cpp:
```diff
@@ -0,0 +1,44 @@
+#include "Traj.h"
```
May I be a stickler about spelling things out?

```suggestion
#include "Trajectory.h"
```

bradleyscollins commented on src/main/include/Traj.h:
```diff
@@ -0,0 +1,11 @@
+
+#include "Constants.h"
+
+double getAngle(const double velocity, const double distance, const double height);
+
+bool isInRange(const double velocity, const double distance, const double angle);
+
+double getDistance(const double ll_angle, const double ll_height, const double target_height);
```
May I also be a stickler about the naming convention in addition to spelling things out? 😬 You might also consider adding units to your measurement parameter names:

```suggestion
double GetAngle(const double velocity_mps, const double distance_meters, const double height_meters);

bool IsInRange(const double velocity_mps, const double distance_meters, const double angle_degrees);

double GetDistance(const double limelight_angle_degrees, const double limelight_height_meters, const double target_height_meters);
```

A potential improvement might be to use the units library, which would enforce the units for you:

```cpp
double GetDistance(const units::angle limelight_angle, const units::length limelight_height, const units::length target_height);
```

Just a thought.

LoganR-FRC-Team34 replied:
I am not personally familiar with the units library and what advantages and capabilities it has, so I'm going to stick with a primitive type for now.

bradleyscollins commented on src/main/include/RobotContainer.h:
```diff
@@ -17,9 +19,13 @@
+    t34::Shooter shooter;
+    t34::Climber climber;
+
+    double arm_angle_setpoint;
```
Suggest either adding the units as a suffix or using the units library:

```suggestion
    double arm_angle_setpoint_degrees;
```
... or ...

```suggestion
    units::angle arm_angle_setpoint;
```

bradleyscollins committed:
3rd party library updates

LoganR-FRC-Team34 committed:
Added controls to switch between percentage-based controls and degree-based controls for the arm. Found an arm scalar. Created the TrajMath class.

LoganR-FRC-Team34 committed:
Merge branch 'LimeLightFunctionality' of https://github.com/Team-34/2024-Crescendo into LimeLightFunctionality

bradleyscollins committed:
Converted GetFiringAngle to equation-as-code

Made the code look a little closer to its reference equation.

Also added a comment with the reference equation and it source.

bradleyscollins committed:
An improvement to the equation-as-code implementation

bradleyscollins commented on src/main/cpp/TrajMath.cpp:
```diff
@@ -0,0 +1,56 @@
+    //   Solving Ballistic Trajectories <https://www.forrestthewoods.com/blog/solving_ballistic_trajectories/>
+    //   See section “Firing Angle to Hit Stationary Target.”
+
+    const auto v² = m_note_max_velocity_mps * m_note_max_velocity_mps;
+    const auto v⁴ = v² * v²;
+    const auto x = m_target_distance_meters;
+    const auto x² = x * x;
+    const auto y = m_target_height_meters;
+    const auto g = 9.80665; // gravity
+
+    const auto gx² = g * x²;
+    const auto v² * y = v²y
+    const auto gx = g * x;
+
+    const auto numerator = v² - sqrt(v⁴ - (g * (gx² + (2  * v²y))));
+    const auto denominator = gx;
+    const auto θ = atan(numerator / denominator);
+
+    return RAD_TO_DEG(θ);
```
Because C++ can handle an expanded Unicode character set in its variable names, it is possible to write code that looks a lot more like the equation it implements. That way, it is a little bit easier to see just how accurately the code implements the source equation.

Have a look at this, and let me know what you think. If you don't care for it, we can revert it.

LoganR-FRC-Team34 replied:
I like it!

bradleyscollins committed:
Fixed typo

bradleyscollins committed:
fix: Corrected identifiers

LoganR-FRC-Team34 committed:
Added limelight_util constructor and Periodic. Added TrajMath constructor. Changed swerve PID values. Added traj_math and limelight_util to Robotcontainer. Added limelight correction controls and binded some states to the D-Pad controls.

LoganR-FRC-Team34 committed:
Merge branch 'LimeLightFunctionality' of https://github.com/Team-34/2024-Crescendo into LimeLightFunctionality

LoganR-FRC-Team34 committed:
WIP

bradleyscollins committed:
Update to WPILib 2024.3.1

bradleyscollins committed:
Tweaked some variable names

bradleyscollins commented on src/main/cpp/TrajMath.cpp:
```diff
@@ -0,0 +1,72 @@
+        (g * m_target_distance_meters) / (pow(m_shooter_angle, 2) * pow(m_note_max_velocity_mps, 2))
+    );
+
+    if (rate > 0)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
```
When returning a Boolean, consider returning the result of the condition rather than writing out a whole `if`-`else` block:

```suggestion
    return rate > 0;
```

Not only do you save some keystrokes; it may even read more clearly.

bradleyscollins commented on src/main/cpp/TrajMath.cpp:
```diff
@@ -0,0 +1,72 @@
+double t34::TrajMath::GetDistanceFromTarget()
+{
+    return (
+        (m_target_height_meters - m_limelight_height_meters) / tan(
+            RAD_TO_DEG( (m_limelight_angle + m_target_ty) )
+        )
```
I think we have it backward here. `tan` [takes radians](https://en.cppreference.com/w/c/numeric/math/tan), not degrees:

```suggestion
        (m_target_height_meters - m_limelight_height_meters) / tan(
            DEG_TO_RAD( (m_limelight_angle + m_target_ty) )
        )
```

That's assuming, by the way, that `m_limelight_angle` and `m_target_ty` are in degrees. If that's correct, it would probably be a good idea to note it in the variable names:

```suggestion
        (m_target_height_meters - m_limelight_height_meters) / tan(
            DEG_TO_RAD( (m_limelight_angle_deg + m_target_ty_deg) )
        )
```

Then again, looking at the name `m_target_ty`, I'm guessing it's actually a lateral distance, not an angle. Are we sure we have the right formula here?

bradleyscollins commented on src/main/cpp/subsystems/LimeLightUtil.cpp:
```diff
@@ -0,0 +1,50 @@
+#include "subsystems/LimeLightUtil.h"
```
Nitpick: Per common practice, the case of the file name should match the case of the class:

```suggestion
#include "subsystems/LimelightUtil.h"
```

I suppose it's possible, by the way, that **did** change the case of the file name on disk, but git didn't pick up the change because Windows path names are not case sensitive. If that's the case, we will need to do some minor wizardry to force it to pick up the change.

bradleyscollins commented on src/main/cpp/subsystems/LimeLightUtil.cpp:
```diff
@@ -0,0 +1,50 @@
+{
+    if (fabs(m_heading_error) > 1.0) {
+
+        if (m_heading_error < 0.0) {
+
+            m_steering_adjust = m_kp*m_heading_error + m_min_command;
+
+        }
+        else 
+        {
+
+            m_steering_adjust = m_kp*m_heading_error - m_min_command;
+      
+    }
```
'Tis a minor thing, but are you familiar with the ternary operator? 

`condition ? value_if_true : value_if_false`

When I see both the `if` and `else` branches setting the same variable, I usually reach for it:

```suggestion
        m_steering_adjust = m_heading_error < 0.0 ?
            m_kp * m_heading_error + m_min_command :
            m_kp * m_heading_error - m_min_command;
```

… or even …

```suggestion
        const auto adjustment = m_heading_error < 0.0 ? m_min_command : -m_min_command;
        m_steering_adjust = m_kp * m_heading_error + adjustment;
```

If either of the above (or some variant thereof) appeals to you, fantastic. If not, feel free to reject.

bradleyscollins commented on src/main/include/TrajMath.h:
```diff
@@ -0,0 +1,43 @@
+        double m_target_distance_meters;
+        double m_target_height_meters;
+        double m_limelight_height_meters;
+        double m_shooter_angle;
+        double m_limelight_angle;
```
Degrees or radians?

bradleyscollins commented on src/main/include/TrajMath.h:
```diff
@@ -0,0 +1,43 @@
+            double note_max_velocity_mps,
+            double target_height_meters,
+            double limelight_height_meters,
+            double shooter_angle,
+            double limelight_angle
```
Degrees or radians?

bradleyscollins committed:
fix: LimelightHelpers update

Fixed liker errors.

LoganR-FRC-Team34 committed:
Upgraded third party libraries

bradleyscollins committed:
Upgraded 3rd-party libraries

LoganR-FRC-Team34 committed:
Addressing pull request comments

LoganR-FRC-Team34 committed:
Merge branch 'LimeLightFunctionality' of https://github.com/Team-34/2024-Crescendo into LimeLightFunctionality

bradleyscollins committed:
Merge branch 'main' into LimeLightFunctionality
